src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/Ordering/OrderAddItemHintReceiver.cs

```diff
@@ -0,0 +1,103 @@
+using Microsoft.Build.Construction;
+using Microsoft.Build.Evaluation;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS.Input.Commands.Ordering
+{
+    [Export(typeof(IProjectChangeHintReceiver))]
+    [Export(typeof(OrderAddItemHintReceiver))]
+    [ProjectChangeHintKind(ProjectChangeFileSystemEntityHint.AddedFileAsString)]
+    [AppliesTo(ProjectCapability.SortByDisplayOrder)]
+    internal class OrderAddItemHintReceiver : IProjectChangeHintReceiver
+    {
+        private readonly IProjectAccessor _accessor;
+
+        private ImmutableHashSet<string> _previousIncludes = ImmutableHashSet<string>.Empty;
+        private OrderAddItemHintReceiverAction _action = OrderAddItemHintReceiverAction.NoOp;
+        private IProjectTree _target = null;
+
+        [ImportingConstructor]
+        public OrderAddItemHintReceiver(IProjectAccessor accessor)
+        {
+            Requires.NotNull(accessor, nameof(accessor));
+
+            _accessor = accessor;
+        }
+
+        public async Task HintedAsync(IImmutableDictionary<Guid, IImmutableSet<IProjectChangeHint>> hints)
+        {
+            if (_action != OrderAddItemHintReceiverAction.NoOp && _target != null && hints.ContainsKey(ProjectChangeFileSystemEntityHint.AddedFile))
+            {
+                var hintSet = hints[ProjectChangeFileSystemEntityHint.AddedFile];
+                foreach (var hint in hintSet)
+                {
+                    var configuredProject = hint.UnconfiguredProject.Services.ActiveConfiguredProjectProvider.ActiveConfiguredProject;
+
+                    await _accessor.OpenProjectForWriteAsync(configuredProject, project =>
+                    {
+                        // We do a sanity re-evaluation to absolutely ensure changes were met.
+                        project.ReevaluateIfNecessary();
+                        var addedElements = GetAddedElements(_previousIncludes, project);
+
+                        switch (_action)
+                        {
+                            case OrderAddItemHintReceiverAction.MoveToTop:
+                                OrderingHelper.TryMoveElementsToTop(project, addedElements, _target);
+                                break;
+                            case OrderAddItemHintReceiverAction.MoveAbove:
+                                OrderingHelper.TryMoveElementsAbove(project, addedElements, _target);
+                                break;
+                            case OrderAddItemHintReceiverAction.MoveBelow:
+                                OrderingHelper.TryMoveElementsBelow(project, addedElements, _target);
+                                break;
+                            default:
+                                break;
+                        }
+                    }).ConfigureAwait(false);
+                }
+            }
+        }
+
+        public async Task HintingAsync(IProjectChangeHint hint)
+        {
+            if (_action != OrderAddItemHintReceiverAction.NoOp)
```

Comment:
Can we tighten the conditions here?

Reply:
What do you mean by 'tighten' in this case?

Reply:
I meant, "make it evaluate to true less often" so that we aren't doing this work when we don't need to.  In the other method there were several conditions, but here there is only the single one for the action.

Reply:
I see. I will put the same conditions as I did in the order method.